signac/common/config.py

```diff
@@ -20,65 +20,64 @@ def _get_project_config_fn(root):
     return os.path.abspath(os.path.join(root, PROJECT_CONFIG_FN))
 
 
-def _get_config_dirname(fn):
-    # We could use pathlib's `.paths` attribute to remove the trailing project
-    # config filename, but that has significant performance
-    # implications: this takes ~200 ns, whereas
-    # os.path.join(*(pathlib.PosixPath(fn).parts[:-2])) takes ~6 us, almost a
-    # 20x slowdown.
-    return fn.replace(os.sep + PROJECT_CONFIG_FN, "")
+def _contains_config_file(root):
+    """Determine if the root directory contains a signac configuration file.
 
+    Parameters
+    ----------
+    root : str
+        Path to search. Uses ``os.getcwd()`` if None (Default value = None).
 
-def _search_local(root):
+    Returns
+    --------
+    str or None
+        The path to the configuration file if one is found, otherwise None.
+    """
     fn_ = _get_project_config_fn(root)
     if os.path.isfile(fn_):
-        yield fn_
+        return root
+    return None
 
 
-def _search_tree(root=None):
-    """Locates signac configuration files in a directory hierarchy.
+def _locate_config_dir(root):
+    """Locates root directory containing a signac configuration file in a directory hierarchy.
 
     Parameters
     ----------
     root : str
-        Path to search. Uses ``os.getcwd()`` if None (Default value = None).
+        Starting path to search.
```

Comment:
Do we need/want to use [`os.path.expanduser`](https://docs.python.org/3/library/os.path.html#os.path.expanduser) or [`os.path.expandvars`](https://docs.python.org/3/library/os.path.html#os.path.expandvars) in any of our path handling? I think the answer is no (users can do that if needed), but I wanted to ask anyway.

Reply:
I'm inclined to agree, let users do that. It's simpler and saves us from incurring the cost unconditionally. However, this point may be moot after we decide what to do on your last comment about whether any of these should be public APIs.